pkg/client/client.go

```diff
@@ -1113,11 +1121,44 @@ func (c *Client) CreateOrUpdateSecret(ctx context.Context, s *v1.Secret) error {
 
 	required := s.DeepCopy()
 	mergeMetadata(&required.ObjectMeta, existing.ObjectMeta)
-
+	// Check if the Secret has an owner reference to a Service, that carries
+	// the annotation with key
+	// service.beta.openshift.io/serving-cert-secret-name and the Secrets
+	// name as the value.
+	// This means that service-ca-operator controls and populates the two
+	// data fields tls.crt and tls.key. We want to retain those on updates
+	// if they existing and are not empty.
+	if c.maybeHasServiceCAData(ctx, required) {
+		if v, ok := existing.Data["tls.crt"]; ok && len(v) > 0 {
+			required.Data["tls.crt"] = v
+		}
+		if v, ok := existing.Data["tls.key"]; ok && len(v) > 0 {
+			required.Data["tls.key"] = v
+		}
+	}
 	_, err = sClient.Update(ctx, required, metav1.UpdateOptions{})
 	return errors.Wrap(err, "updating Secret object failed")
 }
 
+// maybeHasServiceCAData checks if the passed Secret s has at least one owner reference that
+// points to a Service with the annotation service.beta.openshift.io/serving-cert-secret-name: s.name
+func (c *Client) maybeHasServiceCAData(ctx context.Context, s *v1.Secret) bool {
+	for _, owner := range s.OwnerReferences {
+		if owner.Kind != "Service" {
+			continue
+		}
+		sclient := c.kclient.CoreV1().Services(s.GetNamespace())
+		svc, err := sclient.Get(ctx, owner.Name, metav1.GetOptions{})
+		if err != nil {
+			return false
```

Comment:
why not return the error?

Reply:
I wasn't sure how to handle an error here. This current implementation tries to err on the side of caution and remove the tls data if we can't find an owning service for any reason. I suppose it would make sense to check if the error means the service doesn't exist.
I still don't have a clear intuition though how to handle another error. I think the sync loop will probably not even get here if there is another error (lets say we can't communication with the apiserver). And if it does, assuming this data is nothing we care about is still safe. We could also retain the data in that case, though it likely won't make a huge difference.
Actually, now that I think about it, this shouldn't `return` but `continue`. There could be another `OwnerReference` that we can retrieve.